### ISSUE TITLE
User-defined dtype (non-PyObject) in C needing destructors / deallocation?

### ISSUE THREAD

**Issue by EricCousineau-TRI** [numpy/numpy#10721]
Is it possible to define a non-PyObject data type which needs custom element-wise destruction?

We are specifically interested in this for an autodiff scalar, with a dynamic number of gradients -- and also a symbolic expression tree, but I'd assume they're somewhat similar.

We are currently using `pybind11` for our code base, and have used `NPY_OBJECT` for these types, which does the job decently well until we want to have buffer views into our data.

Using a custom dtype seems to permit us to have these custom buffer views, based on tinkering I've done with `numpy.core.test_rational.rational` (using v1.11.0, but not seeing much change elsewhere for `master`).

However, it's not clear to me how I can signal a "cleanup" of an array, e.g. when resizing or when the array goes out of scope. In looking through the code and doc, there are mechanism in place for reference counting object- and record-type arrays, but nothing w.r.t. custom data structures.

Relates https://github.com/RobotLocomotion/drake/issues/8116

**Comment by njsmith** [comment]
I think you're right, and currently there is no way to implement a refcounted/heap-allocated dtype that's not a `PyObject*`, because instead of being handled through a slot in `PyArray_ArrFuncs`, this gets handled by some hard-coded special cases in `array_dealloc` / `PyArray_XDECREF`. Sorry about that.

You might be interested in [this talk](https://www.youtube.com/watch?v=fowHwlpGb34) describing the grants numpy recently got and our plans for them. One of the major goals is to rework the dtype system to make writing these kinds of extensions much easier and more powerful. (I think the grant proposal actually mentioned forward-mode autodifferentiation as one of the motivating examples.) I just added a note about this to our [planning scratch space](https://github.com/njsmith/numpy-grant-planning/blob/master/brainstorming.md), which should probably become something more formal at some point but at least this way it won't get lost.

Unfortunately that doesn't help you much right now – there's a lot of work to do here, and I'd guess we're still at least 6-12 months away from anything landing. But it is on our radar at least, and it's helpful to know that you're interested in this use case :-).

**Comment by EricCousineau-TRI** [comment]
Thanks for the quick response, and for quickly incorporating that into the roadmap! (and sorry for the delay on my part)
I watched the talk, and it looks like some exiting developments! (would be interested in how Numba + Numpy could be leveraged [in pybind11](https://github.com/pybind/pybind11/issues/1268))

I actually just finished prototyping the main behavior that I wanted with `pybind11` + `numpy` user-defined types, and aside from being new the NumPy C API (and battling some `pybind11` internals), it seems like it'll be somewhat usable:
https://github.com/RobotLocomotion/drake/issues/8315#issuecomment-372911602

From what I can see, this issue will not cause a functional difference per se, but it will drain performance. My personal desire is to take a bit of a risk and move forward with the above prototype, and then address the memory leaks when they're a problem (hopefully in 6-12 months :P). (Will see what the rest of the team says.)

If this source of memory leaks crops up as the most pressing bottleneck, I think there may be ways to patch it locally (and in hopefully in an ABI-safe way) pending the upstream feature development, so I don't think we'll be blocked by it just yet.
That being said, I'm looking forward to those features!

**Comment by seberg** [comment]
We have now a clear path to do this via NEP 42.  But it requires a larger refactor of how NumPy traverses arrays for deallocation (that part is clunky and hardcoded to objects and structured arrays containing objects).
The request is still very much valid of course and may happen, but I am not prioritizing it myself quite at this time.
If anyone wants to look at it though, I will definitely help out and discuss how the changes have to look like.

**Comment by EricCousineau-TRI** [comment]
That's awesome! This is also not on my priority queue at the moment, but this is great to see and know about!

**Comment by seberg** [comment]
We actually implemented this now for NEP 42 dtypes, its experimental but its also all there and in use in `numpy-user-dtypes` (the string one).

Closing, since its part of the bigger thing to make it public at this point.  (although I would be interested in seeing usage/experimentation!)
